README.md
# @blazemap/core

[![GitHub license](https://img.shields.io/badge/license-MIT-blue.svg)](https://github.com/patrimart/blazemap-core/blob/main/LICENSE)
[![npm version](https://img.shields.io/npm/v/@blazemap/core.svg?style=flat)](https://www.npmjs.com/package/@blazemap/core)

A blazing fast heatmap library. GPU-accelerated by [GPU.js](https://github.com/gpujs/gpu.js/). Written in TypeScript.

## Installation

### npm

```bash
npm install @blazemap/core --save
```

### yarn

```bash
yarn add @blazemap/core
```

---

## Related Libraries

- [@blazemap/react](#) - coming soon

- [@blazemap/react-leaflet](#) - coming soon

- [@blazemap/react-google](#) - possibly coming soon

---

## How To Use

Basic Blazemap initialization.

```ts
import { blazemap } from '@blazemap/core`;

const blaze = blazemap(canvasElement);

blaze.setPoints([
  [100, 100, 1],
  [145, 130, 1],
]);

blaze.render();
```

Advanced Blazemap initialization.

```ts
import { blazemap, colorsCold, BlazemapOptions } from '@blazemap/core`;

const options: Partial<BlazemapOptions> = {
  radius: 20,
  blur: 10,
  colors: colorsCold,
};

const maxPoints = 2000;

const blaze = blazemap(canvasElement, options, maxPoints);

blaze.setPoints([
  [100, 100, 1],
  [145, 130, 1],
]);

blaze.render();
```

### Resize

Resizing the Blazemap canvas is a performance-heavy operation. It is likely to not exceed 60 FPS (`resize()` and `resizeTo(w, h)`). It's advised to debounce calls to these functions.

### Render

The `render()` function must be called after any changes to the Blazemap. A re-render will easily exceed 60 FPS.

### Destroy

Invoke the `destroy()` function when the `<canvas />` is removed.

---

## Interface

The interface is basically the initializer and the returned functions. Some type aliases are intended to increase clarity.

### `blazemap`

```ts
blazemap(canvas: HTMLCanvasElement, options?: BlazemapOptions, maxPoints?: number): BlazeMap
```

#### Parameters

- `canvas: HTMLCanvasElement` - The canvas on which to render the heatmap.

- `options: Partial<BlazemapOptions>` - (Default: `{ radius: 20, blur: 16, colors: colorsWarm, colorSteps: 0 }`) - Initial options to set.

- `maxPoints: number` - (Default: `1000`) Sets the maximum points to display.

#### Returns

- `addPoint(...points: Points)`: Add one or more points to the heatmap.

- `clearPoints()`: Remove all points from the heatmap.

- `destroy()`: Cleans up the Blazemap resources.

- `modifyPoints(fn: (ps: Points) => Points)`: Accepts a "reducer" to modify the points.

- `render()`: Renders the heatmap to the canvas.

- `resize()`: Resizes the heatmap using the canvas size.

- `resizeTo(w: HexU16, h: HexU16)`: Resizes the heatmap and sets the canvas size.

- `setHeatmap(radius: HexU8, blur: HexU8, colors?: ColorGradient, colorSteps?: HexU8)`: Sets the visual parameters fot the heatmap.

- `setPoints(points: Points)`: Replaces the points in the heatmap.

### `throttle`

Delays invoking `fn` until the next animation frame (or tick).

```ts
throttle<Fn extends (...args: any[]) => any>(fn: Fn) => (...args: Parameters<Fn>) => Promise<ReturnType<Fn>>;
```

---

## Types

```ts
// Float from 0 to 1
export type Proportion = number;

// Unsigned 8-bit Integer
export type HexU8 = number;

// Unsigned 16-bit Integer
export type HexU16 = number;

// Unsigned 32-bit Integer
export type HexU32 = number;

// Defines the color gradient of the heatmap.
export type ColorGradient = Readonly<Record<Proportion, HexU32>>;

// x, y coordinates. The proportional weight of the point 0-1.
export type Point = readonly [x: number, y: number, p: Proportion];

// An array of Points
export type Points = ReadonlyArray<Point>;

// Possible options to set for Blazemap.
export interface BlazemapOptions {
  readonly width: HexU16;
  readonly height: HexU16;
  readonly radius: HexU8;
  readonly blur: HexU8;
  readonly colors: ColorGradient;
  readonly colorSteps: HexU8;
}
```

## Pre-defined Color Gradients

Import a color gradient:

```ts
import { colorsHot } from '@blazemap/core';
```

Available color gradients:

- colorsHot
- colorsWarm
- colorsCold
- colorsGrey
- colorsWhite

---

## Performance

Renders a 1000x1000 canvas with 1000 points in less than 1 ms.

---

## MIT License

### Copyright (c) 2020 Patrick Martin

Permission is hereby granted, free of charge, to any person obtaining a copy
of this software and associated documentation files (the "Software"), to deal
in the Software without restriction, including without limitation the rights
to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
copies of the Software, and to permit persons to whom the Software is
furnished to do so, subject to the following conditions:

The above copyright notice and this permission notice shall be included in all
copies or substantial portions of the Software.

THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN THE
SOFTWARE.


CHANGELOG.md
# Changelog

All notable changes to this project will be documented in this file. See [standard-version](https://github.com/conventional-changelog/standard-version) for commit guidelines.

### [1.1.2](https://github.com/patrimart/blazemap-core/compare/v1.1.1...v1.1.2) (2021-01-14)


### Bug Fixes

* **findMaxWidth:** Improve algorithm ([df161e5](https://github.com/patrimart/blazemap-core/commit/df161e54586babb9e8fe18ff3a510a7e1086662d))

### [1.1.1](https://github.com/patrimart/blazemap-core/compare/v1.1.0...v1.1.1) (2021-01-08)

## [1.1.0](https://github.com/patrimart/blazemap/compare/v1.0.2...v1.1.0) (2020-12-10)


### Features

* **utils:** Add throttle function that delays until the next animation frame. ([6aa8255](https://github.com/patrimart/blazemap/commit/6aa8255c9651bd4a7d378d31aa13b6c3fc32fd69))
* Add color step setting for color gradient. ([057fafc](https://github.com/patrimart/blazemap/commit/057fafc06bae114e97bd04decee272cac4826979))

### [1.0.2](https://github.com/patrimart/blazemap/compare/v1.0.1...v1.0.2) (2020-12-08)

### 1.0.1 (2020-11-26)


src/lib/utils.spec.ts
import test from 'ava';

import { colorsWarm } from './colors';
import * as u from './utils';

test('utilities', async (t) => {
  t.is(u.clampU8(300), 255);
  t.is(u.clampU16(300), 300);
  t.is(u.clampU24(300), 300);
  t.is(u.clampU32(300), 300);
  t.is(u.clampProportion(300), 1);

  t.true(u.pointEq([20, 20, 1], [20, 20, 1]));

  t.is(u.distance([20, 20, 1], [20, 21, 1]), 1);

  t.is(u.normalize(-10, 10)(0), 0.5);

  t.is(u.toHex([255, 255, 255, 255]), 0xffffffff);

  t.deepEqual(u.toRgba(0xffffffff), [255, 255, 255, 255]);

  t.deepEqual(u.tweenRgbas([0, 0, 0, 0], [255, 255, 255, 255], 5), [
    [0, 0, 0, 0],
    [64, 64, 64, 64],
    [128, 128, 128, 128],
    [191, 191, 191, 191],
    [255, 255, 255, 255],
  ]);

  t.true(
    u.genColorScale(
      {
        0: 0x0000ff00,
        0.2: 0x0000ff22,
        0.65: 0x00ff0066,
        1: 0xff0000ee,
      },
      5
    ).length === 256
  );

  t.deepEqual(
    u.DEFAULT_OPTIONS({
      getBoundingClientRect: () => ({ width: 0, height: 0 }),
    } as never),
    {
      width: 0,
      height: 0,
      radius: 20,
      blur: 16,
      colors: colorsWarm,
      colorSteps: 0,
    }
  );

  const OPTIONS = {
    width: 400,
    height: 400,
    radius: 20,
    blur: 16,
    colors: colorsWarm,
    colorSteps: 0,
  };

  t.deepEqual(u.validateOptions(OPTIONS, OPTIONS), OPTIONS);

  t.throws(() => u.validateOptions(OPTIONS, { width: -1 }));

  t.throws(() => u.validateOptions(OPTIONS, { height: -1 }));

  t.throws(() => u.validateOptions(OPTIONS, { radius: -1 }));

  t.throws(() => u.validateOptions(OPTIONS, { blur: -1 }));

  t.throws(() => u.validateOptions(OPTIONS, { colors: { 3: 0 } }));

  await t.notThrowsAsync(async () => {
    const fn = u.throttle((a: number, b: number) => a + b);
    fn(1, 2);
    fn(1, 3);
    const result = await fn(1, 4);
    t.is(result, 3);
  });
});


src/lib/kernel.spec.ts
import test from 'ava';

import { kernel } from './kernel';
import { HexU8, Proportion } from './types';
import { EOP, genColorScale } from './utils';

test('kernel', (t) => {
  const run = () => {
    const points: [x: number, y: number, p: Proportion][] = [
      [100, 100, 1],
      [145, 130, 1],
      [145, 130, 1],
      [110, 150, 1],
      EOP as never,
    ];
    const result: number[] = [];

    function color(r: number): void;
    function color(r: number, g: number): void;
    function color(r: number, g: number, b: number): void;
    function color(r: number, g: number, b: number, a: number): void;
    function color(...rgba: number[]) {
      // console.log(JSON.stringify(rgba));
      result.push(...rgba);
    }

    for (let x = 0; x < 200; x++) {
      for (let y = 0; y < 200; y++) {
        kernel.call(
          {
            output: {
              x: 200,
              y: 200,
              z: 0,
            },
            thread: {
              x,
              y,
              z: 0,
            },
            constants: {
              maxPoints: 1000,
            },
            color,
          },
          points,
          25,
          15,
          (genColorScale({
            0: 0x0000ff00,
            0.2: 0x0000ff22,
            0.65: 0x00ff0066,
            1: 0xff0000ee,
          }) as unknown) as [r: HexU8, g: HexU8, b: HexU8, a: HexU8][],
          2
        );
      }
    }
    return result;
  };

  t.is(run().length, 160_000);
});


src/lib/utils.ts
import { colorsWarm } from './colors';
import {
  BlazemapOptions,
  ColorGradient,
  HexU32,
  HexU8,
  Point,
  RGBa,
} from './types';

export const clamp = (min: number, max: number) => (value: number) =>
  Math.max(Math.min(value, max), min);

export const clampU8 = clamp(0, 0xff);

export const clampU16 = clamp(0, 0xffff);

export const clampU24 = clamp(0, 0xffffff);

export const clampU32 = clamp(0, 0xffffffff);

export const clampProportion = clamp(0, 1);

export const pointEq = (a: Point, b: Point): boolean =>
  a === b || a.every((n, i) => n === b[i]);

export const distance = ([x1, y1]: Point, [x2, y2]: Point) =>
  Math.sqrt(Math.pow(x1 - x2, 2) + Math.pow(y1 - y2, 2));

export const normalize = (min: number, max: number) => (value: number) =>
  (value - min) / (max - min);

export const toRgba = (hex: HexU32): RGBa => {
  if (hex < 0 || hex > 0xffffffff) throw new Error(`Invalid hex value: ${hex}`);
  return [
    (hex * 2 ** -24) | 0,
    (hex & 0x00ff0000) >> 16,
    (hex & 0x0000ff00) >> 8,
    hex & 0x000000ff,
  ];
};

export const toHex = ([r, g, b, a]: RGBa): HexU32 =>
  (r | 0) * 2 ** 24 + ((g | 0) << 16) + ((b | 0) << 8) + (a | 0);

export const tweenRgbas = (
  from: RGBa,
  to: RGBa,
  steps: number
): ReadonlyArray<RGBa> => {
  if (steps < 1)
    throw new Error('tweenRgbas param steps must be greater than 0.');
  const tweens = new Array(steps).fill(from);
  for (let i = 0; i < steps; i++) {
    const delta = normalize(0, steps - 1)(i);
    tweens[i] = [
      Math.round(from[0] + (to[0] - from[0]) * delta),
      Math.round(from[1] + (to[1] - from[1]) * delta),
      Math.round(from[2] + (to[2] - from[2]) * delta),
      Math.round(from[3] + (to[3] - from[3]) * delta),
    ];
  }
  return tweens;
};

export const genColorScale = (
  gradient: ColorGradient,
  colorSteps: HexU8 = 0
): RGBa[] => {
  const [min, max] = Object.keys(gradient)
    .map((k) => {
      const kn = Number(k);
      if (Number.isNaN(kn))
        throw new Error(`ColorGradient keys must be a number: ${gradient}`);
      return kn;
    })
    .reduce(([m, M], v) => [Math.min(m, v), Math.max(M, v)], [0xffffffff, 0]);

  const [from, to, ...tail] = Object.entries(gradient)
    .map(([k, v]) => [normalize(min, max)(Number(k)), toRgba(v)] as const)
    .sort((a, b) => a[0] - b[0]);

  const ranges: ReadonlyArray<
    readonly [from: readonly [number, RGBa], to: readonly [number, RGBa]]
  > = tail
    .reduce(([[f, t], ...tl], g) => [[t, g] as const, [f, t] as const, ...tl], [
      [from, to] as const,
    ])
    .reverse();

  const scale: RGBa[] = new Array(256).fill(0);
  for (let i = 0; i < ranges.length; i++) {
    const [[mn, f], [mx, t]] = ranges[i];
    const index = Math.round(256 * mn);
    const steps = Math.round(256 * mx - index);
    tweenRgbas(f, t, steps).forEach((r, i) => (scale[i + index] = r));
  }

  if (colorSteps !== 0) {
    for (let i = 254; i > 0; i--) {
      if (i % Math.round(255 / colorSteps) === 0) {
        i--;
      } else {
        scale[i] = scale[i + 1];
      }
    }
  }

  return scale;
};

export const EOP: Point = [0, 0, 10_000];

export const DEFAULT_OPTIONS = (
  canvas: HTMLCanvasElement
): BlazemapOptions => ({
  width: canvas.getBoundingClientRect().width,
  height: canvas.getBoundingClientRect().height,
  radius: 20,
  blur: 16,
  colors: colorsWarm,
  colorSteps: 0,
});

export function assertValidNumber(
  v: unknown,
  min = 1,
  max = 65_535
): asserts v is number {
  if (!(typeof v === 'number' && v >= min && v <= max)) {
    throw new Error(`The value is not a number within ${min} and ${max}: ${v}`);
  }
}

export const validateOptions = (
  { width, height, radius, blur, colors, colorSteps }: BlazemapOptions,
  options: Readonly<Partial<BlazemapOptions>> = {}
): BlazemapOptions => {
  const opts = { width, height, radius, blur, colors, colorSteps };

  if (options.width) {
    assertValidNumber(options.width);
    opts.width = options.width;
  }

  if (options.height) {
    assertValidNumber(options.height);
    opts.height = options.height;
  }

  if (options.radius) {
    assertValidNumber(options.radius, 1, 255);
    opts.radius = options.radius;
  }

  if (options.blur) {
    assertValidNumber(options.blur, 0, 255);
    opts.blur = options.blur;
  }

  if (options.colors) {
    Object.entries(options.colors).forEach(([k, v]) => {
      assertValidNumber(Number(k), 0, 1);
      assertValidNumber(v, 0, 0xffffffff);
    });
    opts.colors = options.colors;
  }

  if (options.colorSteps) {
    assertValidNumber(options.colorSteps, 0, 255);
    opts.colorSteps = options.colorSteps;
  }

  return opts;
};

/**
 * Delays invocation until the next animation frame (or tick).
 */
export const throttle = <Fn extends (...args: any[]) => any>(fn: Fn) => {
  let resp: Promise<ReturnType<Fn>> | null = null;
  return (...args: Parameters<Fn>) => {
    if (resp === null) {
      resp = new Promise((resolve) =>
        (globalThis?.requestAnimationFrame ?? globalThis.setTimeout)(() => {
          resp = null;
          resolve(fn(...args));
        })
      );
    }
    return resp;
  };
};


src/lib/kernel.ts
import { IKernelFunctionThis } from 'gpu.js';

import { HexU8, Proportion } from './types';

export type KernelInit = (
  this: IKernelFunctionThis<{ maxPoints: number }>,
  points: [x: number, y: number, p: Proportion][],
  radius: number,
  blur: number,
  colorScale: [r: HexU8, g: HexU8, b: HexU8, a: HexU8][],
  maxWeight: number
) => void;

export function kernel(
  this: IKernelFunctionThis<{ maxPoints: number }>,
  points: [x: number, y: number, p: Proportion][],
  radius: number,
  blur: number,
  colorScale: [r: HexU8, g: HexU8, b: HexU8, a: HexU8][],
  maxWeight: number
) {
  function clampit(v: number) {
    return Math.round(Math.max(Math.min(v, 255), 0));
  }
  function distance(x1: number, y1: number, x2: number, y2: number) {
    return Math.sqrt(Math.pow(x1 - x2, 2) + Math.pow(y1 - y2, 2));
  }
  function norm(v: number): number {
    return (v - 0) / (255 - 0);
  }
  function easeFade(d: number, r: number, b: number) {
    if (d < r - b * 0.5) return 1;
    return (
      1 -
      Math.min(
        1,
        (((r + b * 0.5) / (r + b * 0.5 - (r - b * 0.5))) *
          (d - (r - b * 0.5))) /
          (r + b * 0.5)
      )
    );
  }

  const x = this.thread.x;
  const y = this.thread.y;

  let weight = 0;

  for (let i = 0; i < this.constants.maxPoints; i++) {
    if (points[i][2] == 10_000) break;
    const d = distance(points[i][0], this.output.y - points[i][1], x, y);
    const w = easeFade(d, radius, blur);
    weight += points[i][2] * w;
  }

  const wc = clampit((weight / maxWeight) * 255);

  this.color(
    norm(colorScale[wc][0]),
    norm(colorScale[wc][1]),
    norm(colorScale[wc][2]),
    norm(colorScale[wc][3])
  );
}

export const kernelInit = (new Function(
  'points',
  'radius',
  'blur',
  'colorScale',
  'maxWeight',
  `{
    function clampit(v) {
      return Math.round(Math.max(Math.min(v, 255), 0));
    }
    function distance(x1, y1, x2, y2) {
        return Math.sqrt(Math.pow(x1 - x2, 2) + Math.pow(y1 - y2, 2));
    }
    function norm(v) {
        return (v - 0) / (255 - 0);
    }
    function easeFade(d, r, b) {
        if (d < r - b * 0.5)
            return 1;
        return (1 -
            Math.min(1, (((r + b * 0.5) / (r + b * 0.5 - (r - b * 0.5))) *
                (d - (r - b * 0.5))) /
                (r + b * 0.5)));
    }
    var x = this.thread.x;
    var y = this.thread.y;
    var weight = 0;
    for (var i = 0; i < this.constants.maxPoints; i++) {
        if (points[i][2] == 10000)
            break;
        var d = distance(points[i][0], this.output.y - points[i][1], x, y);
        var w = easeFade(d, radius, blur);
        weight += points[i][2] * w;
    }
    var wc = clampit((weight / maxWeight) * 255);
    this.color(norm(colorScale[wc][0]), norm(colorScale[wc][1]), norm(colorScale[wc][2]), norm(colorScale[wc][3]));
  }`
) as unknown) as KernelInit;


src/lib/blazemap.ts
import { GPU, IKernelRunShortcut } from 'gpu.js';

import { kernelInit } from './kernel';
import { BlazemapOptions, ColorGradient, HexU8, Point, Points } from './types';
import {
  assertValidNumber,
  DEFAULT_OPTIONS,
  EOP,
  genColorScale,
  validateOptions,
} from './utils';

/**
 * Initialize the Blazemap.
 * @param canvas
 * @param options
 * @param maxPoints
 */
export const blazemap = (
  canvas: HTMLCanvasElement,
  options: Readonly<
    Partial<Pick<BlazemapOptions, 'radius' | 'blur' | 'colors' | 'colorSteps'>>
  > = {},
  maxPoints = 1000
) => {
  assertValidNumber(maxPoints, 1, 100_000);
  let opts = validateOptions(DEFAULT_OPTIONS(canvas), options);
  let colorScale = genColorScale(opts.colors, opts.colorSteps);
  let pts: Point[] = [];
  let maxWeight = 1000;
  let currentKernel: IKernelRunShortcut | undefined;

  const context = canvas.getContext('webgl2', { premultipliedAlpha: false });
  if (context === null) throw new Error('Canvas context returned null.');

  const gpu = new GPU({ canvas, context });

  const createKernel = () => {
    currentKernel?.destroy();
    currentKernel = gpu
      .createKernel(kernelInit, {
        argumentTypes: {
          a: 'Array',
          b: 'Integer',
          c: 'Integer',
          d: 'Array',
          e: 'Integer',
        },
        graphical: true,
        immutable: true,
        dynamicArguments: true,
        dynamicOutput: true,
        constants: { maxPoints },
        output: [opts.width, opts.height],
      })
      .setLoopMaxIterations(maxPoints);
  };

  createKernel();

  const findMaxWeight = () => {
    const diam = opts.radius - opts.blur * 0.5;
    const grid = [1];

    for (let i = 0; i < pts.length; i++) {
      const [x, y, p] = pts[i];
      const index = Math.round((x / diam) * (y / diam));
      grid[index] = (grid[index] ?? 0) + p;
    }

    maxWeight = grid.reduce((max, v) => Math.max(max, v));
  };

  /****************************************************************************
   * PUBLIC FUNCTIONS
   */

  const setHeatmap = (
    radius: number,
    blur: number,
    colors?: ColorGradient,
    colorSteps?: HexU8
  ) => {
    opts = validateOptions(opts, { radius, blur, colors, colorSteps });
    colorScale = genColorScale(opts.colors, opts.colorSteps);
    findMaxWeight();
  };

  const resize = () => {
    opts = validateOptions(opts, {
      width: canvas.width,
      height: canvas.height,
    });
    findMaxWeight();
    createKernel();
  };

  const resizeTo = (width: number, height: number) => {
    opts = validateOptions(opts, { width, height });
    canvas.width = opts.width;
    canvas.height = opts.height;
    findMaxWeight();
    createKernel();
  };

  const setPoints = (points: Points) => {
    pts = points.filter((p) => p[0] > 0 && p[1] > 0);
    findMaxWeight();
  };

  const addPoint = (...point: Points) => {
    setPoints(point.concat(pts));
  };

  const modifyPoints = (fn: (ps: Points) => Points) =>
    setPoints(fn(pts.slice(0)));

  const clearPoints = () => setPoints([]);

  const render = () => {
    currentKernel?.(
      (pts.concat([EOP]) as unknown) as number[][],
      opts.radius,
      opts.blur,
      (colorScale as unknown) as number[][],
      maxWeight
    );
  };

  const destroy = () => gpu.destroy();

  return {
    render,
    resize,
    resizeTo,
    setHeatmap,
    setPoints,
    addPoint,
    modifyPoints,
    clearPoints,
    destroy,
  };
};

export type BlazeMap = ReturnType<typeof blazemap>;
